modern-publish-plugin.md

```diff
@@ -0,0 +1,211 @@
+# Modern Gradle Publish Plugin
+
+
+## Goals
+
+- less magic configuration, more explicit APIs
+- provide APIs for everything represented through properties
+- remove or deprecate APIs that are already provided by Gradle
+- continue providing a way to get started with minimal configuration
+
+## New Structure
+
+For compatibility purposes the current `com.vanniktech.maven.publish` plugin
+behavior is kept as it is, with the exception of APIs that we decide to remove. 
+Most of the goals are solved through a set of new plugins for specific 
+platforms. These plugins provide APIs that replace the magic setup mechanisms
+of the current plugin. To avoid having to maintain two implentations the old
+plugin will internally just apply the new plugins and use their public APIs. 
+Over time we can decide if `com.vanniktech.maven.publish` should be deprecated
+or removed.
+
+### Plugins
+
+Specific plugins for each supported project type:
+- `com.vanniktech.maven.publish.java`
+- `com.vanniktech.maven.publish.gradle`
+- `com.vanniktech.maven.publish.android`
+- `com.vanniktech.maven.publish.kotlin.js`
+- `com.vanniktech.maven.publish.kotlin.mpp`
+
+There are two reasons for going with individual plugins. 
+
+
+On top of that there will be a `com.vanniktech.maven.publish.base` plugin that 
+gets applied by all of these and the existing `com.vanniktech.maven.publish` as 
+described above.
+
+#### `com.vanniktech.maven.publish.base`
+
+- applies `maven-publish` Gradle plugin
+- provides the API that is shared for all project types
+- can be used for shared configurations using `plugins.withId(...)` in projects with mixed types
+
+```
+mavenPublish {
+  # applies signing plugin and tells Gradle to sign all publications
+  signAllPublications()
+  # adds the Sonatype release/snapshot repo as targets based on version
+  # uses Gradle credentials API for username/password
+  # automatically calls signAllPublications()
+  # sets nexus.baseUrl
+  publishToMavenCentral()
+
+  # might not be needed as it defaults to the mutable `project.getGroup()`
+  groupId = "com.example"
+  # might not be needed as it defaults to the mutable `project.getGroup()`
+  version = "1.0.0"
+  # Base because their might be multiple artifacts produced by one module.
+  # Without this the arifact id would always be based on `project.getName()`.
+  # Implementation in Kotlin MPP is kind of hacky using replace so it might be
+  # best to just not support it. How do we handle compatibility?
+  baseArtifactId = "library"
```

Comment:
Should we have `groupId`, `version` and `baseArtifactId` extension properties? 
Gradle itself has an API to set/change them on a publication, so it's not always limited to the project name, but it gets hacky with publications not created by us.

Reply:
How about we start without them and if required, we add them?